src/symbolic_dimensions.jl

```diff
@@ -35,49 +35,59 @@ to one which uses `Dimensions` as its dimensions (i.e., base SI units)
-Base.getindex(d::SymbolicDimensions{R}, k::Symbol) where {R} = getproperty(d, k)
+function Base.getproperty(d::SymbolicDimensions{R}, s::Symbol) where {R}
+    nzdims = getfield(d, :nzdims)
+    i = get(ALL_MAPPING, s, INDEX_TYPE(0))
+    iszero(i) && error("$s is not available as a symbol in SymbolicDimensions. Symbols available: $(ALL_SYMBOLS).")
+    ii = searchsortedfirst(nzdims, i)
+    if ii <= length(nzdims) && nzdims[ii] == i
+        return getfield(d, :nzvals)[ii]
+    else
+        return zero(R)
+    end
+end
+Base.propertynames(::SymbolicDimensions) = ALL_SYMBOLS
+Base.getindex(d::SymbolicDimensions, k::Symbol) = getproperty(d, k)
 constructor_of(::Type{<:SymbolicDimensions}) = SymbolicDimensions
 
-SymbolicDimensions{R}(d::SymbolicDimensions) where {R} = SymbolicDimensions{R}(data(d))
-(::Type{D})(; kws...) where {D<:SymbolicDimensions} = D(DEFAULT_DIM_BASE_TYPE; kws...)
-(::Type{D})(::Type{R}; kws...) where {R,D<:SymbolicDimensions} =
-    let constructor=constructor_of(D){R}
-        length(kws) == 0 && return constructor(SA.spzeros(R, length(ALL_SYMBOLS)))
-        I = [ALL_MAPPING[s] for s in keys(kws)]
-        V = [tryrationalize(R, v) for v in values(kws)]
-        data = SA.sparsevec(I, V, length(ALL_SYMBOLS))
-        return constructor(data)
+SymbolicDimensions{R}(d::SymbolicDimensions) where {R} = SymbolicDimensions{R}(getfield(d, :nzdims), convert(Vector{R}, getfield(d, :nzvals)))
+SymbolicDimensions(; kws...) = SymbolicDimensions{DEFAULT_DIM_BASE_TYPE}(; kws...)
+function SymbolicDimensions{R}(; kws...) where {R}
+    if isempty(kws)
+        return SymbolicDimensions{R}(Vector{INDEX_TYPE}(undef, 0), Vector{R}(undef, 0))
     end
+    I = INDEX_TYPE[ALL_MAPPING[s] for s in keys(kws)]
+    p = sortperm(I)
+    V = R[tryrationalize(R, kws[i]) for i in p]
+    return SymbolicDimensions{R}(permute!(I, p), V)
+end
+(::Type{<:SymbolicDimensions})(::Type{R}; kws...) where {R} = SymbolicDimensions{R}(; kws...)
 
-function Base.convert(::Type{Qout}, q::Quantity{<:Any,<:Dimensions}) where {T,D<:SymbolicDimensions,Qout<:Quantity{T,D}}
-    output = Qout(
-        convert(T, ustrip(q)),
-        D;
-        m=ulength(q),
-        kg=umass(q),
-        s=utime(q),
-        A=ucurrent(q),
-        K=utemperature(q),
-        cd=uluminosity(q),
-        mol=uamount(q),
-    )
-    SA.dropzeros!(data(dimension(output)))
-    return output
+function Base.convert(::Type{Quantity{T,SymbolicDimensions}}, q::Quantity{<:Any,<:Dimensions}) where {T}
+    return convert(Quantity{T,SymbolicDimensions{DEFAULT_DIM_BASE_TYPE}}, q)
+end
+function Base.convert(::Type{Quantity{T,SymbolicDimensions{R}}}, q::Quantity{<:Any,<:Dimensions}) where {T,R}
+    syms = (:m, :kg, :s, :A, :K, :cd, :mol)
+    vals = (ulength(q), umass(q), utime(q), ucurrent(q), utemperature(q), uluminosity(q), uamount(q))
+    I = INDEX_TYPE[ALL_MAPPING[s] for (s, v) in zip(syms, vals) if !iszero(v)]
```

Comment:
I think this `if !iszero(v)` is what is introducing the second bug. I'll push a fix for this. For safety I think this `convert` should just call the normal constructor instead of writing out the `sortperm` stuff again.

Reply:
On purpose, I did not use the same constructor since this will lead to unnecessary zero dimensions. Using the same constructor will also change behaviour compared with the master branch.

So my suggestion is to revert the last commit (commits?) and instead fix and test the bug in the function.

Reply:
Got it, thanks!